dispatch/BESContainerStorageList.cc

```diff
@@ -49,371 +51,205 @@ using std::endl;
 using std::string;
 using std::ostream;
 
-BESContainerStorageList *BESContainerStorageList::d_instance = nullptr;
-static std::once_flag d_euc_init_once;
-
-BESContainerStorageList::BESContainerStorageList() :
-        _first(0)
-{
-}
-
-BESContainerStorageList::~BESContainerStorageList()
-{
-    BESContainerStorageList::persistence_list *pl = _first;
-    while (pl) {
-        if (pl->_persistence_obj) {
-            delete pl->_persistence_obj;
-        }
-        BESContainerStorageList::persistence_list *next = pl->_next;
-        delete pl;
-        pl = next;
-    }
-}
-
-/** @brief Add a persistent store to the list
+/**
+ * @brief Add a persistent store pointer to the list.
  *
- * Each persistent store has a name. If a persistent store already exists in
- * the list with that name then the persistent store is not added. Otherwise
- * the store is added to the list.
+ * Each persistent store has a name (retrieved via getPersistenceName()).
+ * If a persistent store pointer is already in the list with that name,
+ * the new pointer is not added. Otherwise, the pointer is added to the
+ * end of the list with a reference count of 1.
  *
+ * The list does NOT take ownership of the pointer 'cp'. Caller manages lifetime.
  * The persistent stores are searched in the order in which they were added.
  *
- * @param cp persistent store to add to the list
- * @return true if successfully added, false otherwise
+ * @param cp Non-owning pointer to the persistent store to add to the list.
+ * @return True if successfully added, false otherwise (e.g., cp is null,
+ * or a store with the same name already exists).
  * @see BESContainerStorage
+ * @see BESContainerStorageList::StorageEntry
  */
 bool BESContainerStorageList::add_persistence(BESContainerStorage *cp)
 {
-    std::lock_guard<std::recursive_mutex> lock_me(d_cache_lock_mutex);
-
-    bool ret = false;
-    if (!_first) {
-        _first = new BESContainerStorageList::persistence_list;
-        _first->_persistence_obj = cp;
-        _first->_reference = 1;
-        _first->_next = 0;
-        ret = true;
+    std::lock_guard<std::recursive_mutex> lock(d_cache_lock_mutex);
+
+    if (!cp) {
+        return false;
     }
-    else {
-        BESContainerStorageList::persistence_list *pl = _first;
-        bool done = false;
-        while (done == false) {
-            if (pl->_persistence_obj->get_name() != cp->get_name()) {
-                if (pl->_next) {
-                    pl = pl->_next;
-                }
-                else {
-                    pl->_next = new BESContainerStorageList::persistence_list;
-                    pl->_next->_reference = 1;
-                    pl->_next->_persistence_obj = cp;
-                    pl->_next->_next = 0;
-                    done = true;
-                    ret = true;
-                }
-            }
-            else {
-                done = true;
-                ret = false;
-            }
-        }
+
```

Comment:
I am not familiar with this part of code. It looks like you are doing the right thing. So does this also cause ASAN complain?

Reply:
Yes, there was a memory leak in the broken implementation of the singleton pattern.